src/allocator.rs

```diff
@@ -202,33 +286,69 @@ impl Allocator {
     }
 
     pub fn new_substr(&mut self, node: NodePtr, start: u32, end: u32) -> Result<NodePtr, EvalErr> {
-        if self.atom_vec.len() == MAX_NUM_ATOMS {
+        if self.atom_vec.len() + self.small_atoms == MAX_NUM_ATOMS {
             return err(self.nil(), "too many atoms");
         }
-        let (ObjectType::Bytes, idx) = node.node_type() else {
-            return err(node, "(internal error) substr expected atom, got pair");
-        };
-        let atom = self.atom_vec[idx as usize];
-        let atom_len = atom.end - atom.start;
-        if start > atom_len {
-            return err(node, "substr start out of bounds");
-        }
-        if end > atom_len {
-            return err(node, "substr end out of bounds");
-        }
-        if end < start {
-            return err(node, "substr invalid bounds");
+        match node.node_type() {
+            (ObjectType::Pair, _) => err(node, "(internal error) substr expected atom, got pair"),
+            (ObjectType::Bytes, idx) => {
+                let atom = self.atom_vec[idx as usize];
+                let atom_len = atom.end - atom.start;
+                if start > atom_len {
+                    return err(node, "substr start out of bounds");
+                }
+                if end > atom_len {
+                    return err(node, "substr end out of bounds");
+                }
+                if end < start {
+                    return err(node, "substr invalid bounds");
+                }
+                let idx = self.atom_vec.len();
+                self.atom_vec.push(AtomBuf {
+                    start: atom.start + start,
+                    end: atom.start + end,
+                });
+                Ok(NodePtr::new(ObjectType::Bytes, idx))
+            }
+            (ObjectType::SmallAtom, val) => {
+                let len = len_for_value(val) as u32;
+                if start > len {
+                    return err(node, "substr start out of bounds");
+                }
+                if end > len {
+                    return err(node, "substr end out of bounds");
+                }
+                if end < start {
+                    return err(node, "substr invalid bounds");
+                }
+                let buf: [u8; 4] = val.to_be_bytes();
+                let buf = &buf[4 - len as usize..];
+                let substr = &buf[start as usize..end as usize];
+                if !canonical_positive_integer(substr) {
```

Comment:
I wonder if this whole block could be factored elsewhere. It seems hard to believe this would be the only place with this logic.